23.11 ZHF Hackathon and UX Workshop report (#4)
nix-community/ZurichZHF: 2023-11-report/index.md

# Zürich 23.11 ZHF Hackathon and UX Workshop report

On the last weekend of November 2023, the Zürich Friends of Nix hosted a multi-event around the [Zero Hydra Failures release preparations for NixOS 23.11](https://discourse.nixos.org/t/lets-have-a-great-23-11-release-cycle/33765).
It featured a two-day hackathon and a UX workshop, that all took place at [Eastern Switzerland University of Applied Sciences](https://www.ost.ch/) (OST) right next to Lake Zürich.

# Background and preparations

The Nix ecosystem is growing, and so is the community building it.
This can also be felt in the now regular Zürich ZHF hackathon, which has seen more and more participants over time.

> The event is designed to encourage people to learn to know each other, hack on everything Nix, and together shape the future of open source.

Reviewer: ```suggestion
The event is designed to encourage people to get to know each other, hack on everything Nix, and shape the future of open source together.
```

> The third iteration again incorporated participant feedback from last time, and also experience from in-person [workshops at NixCon 2023](https://discourse.nixos.org/t/nixcon-governance-workshop/32705):

Reviewer: ```suggestion
On this third iteration, we incorporated feedback from last time as well as experiences from [NixCon 2023 in-person workshops](https://discourse.nixos.org/t/nixcon-governance-workshop/32705):
```

- More time for hacking and mingling
- Shorter workshop sessions
- A variety of brief talks for overview and inspiration
- Increased coffee throughput

Preparations started in June 2023, right after the event, by writing up a report, processing feedback, and sketching a schedule for the next event.
In the four months before the event, we invited many guests and followed up with reminders, refined the agenda, and prepared some of the logistics.

> While originally there were ideas to make some progress on project governance topics, some key people couldn't make it this time and it turned out that the schedule wouldn't have allowed for it anyway.

Reviewer: ```suggestion
While originally there were plans to work on project governance topics, some key people couldn't make it to the workshop this time and it turned out that the schedule wouldn't have allowed for it anyway.
```

> A brief survey on state the 6-months objectives from the last workshop showed that it wouldn't make sense to have a separate agenda item for that.

Reviewer: ```suggestion
A brief survey on the 6-month-objectives from the last workshop showed that there was no desire to put this topic on our agenda.
```

Author: ```suggestion
A brief survey of the 6-month-objectives from the last workshop showed that it wouldn't make sense to have a separate agenda item for those.
```

No, it quite literally didn't make sense because there was so little progress. We can't tell if there was or wasn't desire other than our own.

There is progress on a few of the objectives though:
- @zimbatm launched [Nix packaging as a service](https://discourse.nixos.org/t/beta-launch-nix-packaging-as-a-service/34683)
- The [commercial support page](https://nixos.org/community/commercial-support) on nixos.org got more specific with respect to which support services are offered
- @jeremiahs is developing a plan to implement a [CIS benchmark](https://www.cisecurity.org/cis-benchmarks) and configuration profile for NixOS
- [RFC 140 got merged](https://media.ccc.de/v/nixcon-2023-35713-not-all-packages-anymore-nix) and @infinisil worked on [better tooling to work with local files in derivations](https://www.tweag.io/blog/2023-11-28-file-sets/)

# Event

The event took place in two large rooms and the hallway, and spanned two days, starting Saturday at 10:00 and ending at 19:00 with a group photo, and continuing on Sunday at 10:00 until it phased out around 17:00.

![](./XXX.jpg)

We had around 35 attendees on Saturday and 25 on Sunday, arriving from all over Europe, and we were happy to meet both first-time participants as well as regulars.

> In addition to all the people giving presentations, we were particularly happy welcome NixOS Foundation board member @zimbatm, @delroth from the infrastructure team, `niv` author @nmattia, and many others.

Reviewer: ```suggestion
We had around 35 attendees on Saturday and 25 on Sunday, arriving from all over Europe, and we were glad to meet both first-time participants as well as regulars.
In addition to all the people giving presentations, we were particularly happy to welcome NixOS Foundation board member @zimbatm, @delroth from the infrastructure team, `niv` author @nmattia, and many others.
```

# Lightning talks

At the beginning of each day we had guests highlighting recent developments in the ecosystem in brief talks.

On Saturday:
- Alex (@ners) welcomed everyone and provided an overview of the two following days
- Valentin (@fricklerhandwerk) gave a brief update on recent work in Nix maintenance and documentation
- Elton (@EltonUX) summarised the last two workshop sessions ([November 2022](https://discourse.nixos.org/t/2022-11-26-nix-nixpkgs-nixos-hackathon-in-zurich-rapperswil-jona/23402/11) and [May 2023](https://discourse.nixos.org/t/zurich-23-05-zhf-hackathon-and-workshop-report/29093)), and gave taste of what would expect participants this time
- Ivan (@iminicik) and Auguste (@augustebaum) reported on their mob programming experience with Summer of Nix
- Silvan (@infinisil) made his traditional introduction to working on ZHF

> On Sunday:

Reviewer: Missing a proper line break between the days.

```suggestion
On Saturday:
- Alex (@ners) welcomed everyone and provided an overview of the two following days
- Valentin (@fricklerhandwerk) gave a brief update on recent work in Nix maintenance and documentation
- Elton (@EltonUX) summarised the last two workshop sessions ([November 2022](https://discourse.nixos.org/t/2022-11-26-nix-nixpkgs-nixos-hackathon-in-zurich-rapperswil-jona/23402/11) and [May 2023](https://discourse.nixos.org/t/zurich-23-05-zhf-hackathon-and-workshop-report/29093)), and gave taste of what would expect participants this time
- Ivan (@iminicik) and Auguste (@augustebaum) reported on their mob programming experience with Summer of Nix
- Silvan (@infinisil) made his traditional introduction to working on ZHF

On Sunday:
```

- Ivan (@imincik) introduced the [Nixpkgs geospatial software maintainer team](https://discourse.nixos.org/t/creating-nix-geospatial-team/23454) he founded
- Matthias (@mat) brought our attention to how commercial interests impact community dynamics in open source

# ZHF Hackathon

Before and after the talks in the morning, there was plenty of time for introductions or catching up.

> Plenty of Coffee, fresh vegan sandwiches and wraps, and λ shaped cookies were waiting in the hallway.

Reviewer: ```suggestion
Plenty of coffee, fresh vegan sandwiches and wraps, and λ shaped cookies were waiting in the hallway.
```

Snacking was accompanied by hacking, in which most of the attendees participated this time.
@infinisil was available to assist beginners and experts alike with navigation and debugging.
People worked on a range of issues, from trivial typos, over cursed Python environments, to surprisingly deep issues with the Nixpkgs `stdenv` infrastructure and hard-coded limitations in Hydra.
Over the weekend, [22 ZHF pull requests](https://github.com/NixOS/nixpkgs/issues?q=ZurichZHF+created%3A%3E2023-11-24) were opened and most of them got merged.

# UX workshop

This was the third iteration of the Nix UX workshop led by Elton Vecchietti, Director of UX at Modus Create.

In the first event a year ago, the group had outlined a typical user journey, tracing interactions across a sequence of touchpoints.

> In the second event half a year ago, we made comparisons to similar tools, framed processes in the ecosystem in terms of resource flows, and tried to distill central value propositions.

Reviewer: ```suggestion
In the second event half a year ago, we compared Nix to similar tools, framed processes in the ecosystem in terms of resource flows, and tried to distill central value propositions.
```

This time, the goal was to come up with an information architecture for the central hub of the ecosystem, with the intention of supporting the Nix marketing team in their ongoing effort to rework nixos.org.

> In three sessions distributed over two days, guided by Elton and building on the understanding gathered so far, a group of 5-10 people tried to systematically answer three major questions:

Reviewer: ```suggestion
In three sessions spread out over two days, guided by Elton and building on the understanding gathered so far, a group of 5-10 people tried to systematically answer three major questions:
```

- Academic researcher
- Employee developer
- DevOps engineer
- Sales executive
- ...

Each of the labels got a verbal explanation that was confirmed by the group, sometimes accompanied by a written comment on the digital sticky notes.

![](./1_user_types.png)

After that, participants were asked to cluster these user types.
When we came back to this panel to map these clusters to the objects of interest found in the second assignment, they were also labeled:
- Enthusiasts
- Explorers
- Pragmatists
- Contributors
- Decisionmakers

## Objects of interest

The second assignment was to determine objects of interaction in the ecosystem, with the ultimate goal to provide short and obvious navigation paths to the most important ones in the information architecture we were working towards.

The group struggled a bit with untangling objects from their properties and acting on the objects, but eventually came up with items such as:
- Package
- Option search
- Example configuration
- Development environment
- Build instructions
- Tutorial
- Command-line interface
- Source code
- Brand identity
- ...

Again, each item was explained verbally and sometimes annotated or reworded for clarity.

> The collection was again clustered, and also got preliminary priorities with labels `{1,2,3}{A,B,C}` denoting three top-level categories 1-3, each with three items A-C:

Reviewer: Even our group struggled to understand what exactly we were doing at this point. So this is too complex for anyone that wasn't there. I'd try to simplify, maybe something like this.

```suggestion
The collection was then clustered by topic (letters A-C) and prioritised (numbers 1-3):
```

At this point, I'm not even sure _that's_ correct anymore. 😅

Then we tried to find the three most important user types and the objects they interact with, which revealed the overlap and hints at which objects may end up to be more prominent in the information architecture.

![](./3_user_types.png)

## Topics

On the second workshop day, with a smaller group due to participants' time constraints (and also some exhaustion from the first two sessions), we re-evaluated the collections of user types and objects and were tasked to find 9 topics and refine the objects of interests that belong to these topics, but without a particular order.

The following topics were identified:
- Downloads
- Use cases
- Packages
- Customisation
- What does Nix do?
- Learning
- Documentation
- Community
- Identity

The result was presented to the group of ZHF hackers in the other room for a quick validation, which added a few items to the overview.

![](./4_topics.png)

## Summary

While at times it felt like hair splitting over terms and definitions, and the fast pace necessitated glimpsing over some inaccuracies, the procedure was highly organised and produced new insights for participants.

> One of them is how Nix is at the nexus between different ecosystems, wrapping artifacts of some (e.g. software builds, Git or Vim configurations, service definitions) and producing artifacts for others (e.g. Unix user environments, Linux distributions, Docker images, virtual machine images) -- anything that can be expressed in terms of files in a file system.

Reviewer: Not sure what this means/adds.

```suggestion
One of them is how Nix is at the nexus between different ecosystems, wrapping artifacts of some (e.g. software builds, Git or Vim configurations, service definitions) and producing artifacts for others (e.g. Unix user environments, Linux distributions, Docker images, virtual machine images).
```

> Elton pointed out that the kind of process we went through in three weekends usually takes on the order of six weeks and involves all stakeholders of an organisation, so we can consider it as a glimpse at what's possible.

Reviewer: ```suggestion
Elton pointed out that the kind of process we went through in three weekends usually takes somewhere around six weeks and involves all stakeholders of an organisation, so we can consider it as a glimpse of what's possible.
```

Therefore it's not surprising that these and the past sessions really felt like work, and that with Elton's guidance and in a short amount of time we still have achieved a great increase of clarity that was otherwise unavailable.

> Some of the topics, user types, and objects of interest we discussed and compiled are already represented on nixos.org, nix.dev, and Discourse, with varying degrees of structure.

Reviewer: Maybe it makes sense to name a specific example here?

Author: ```suggestion
Some of the topics, user types, and objects of interest we discussed and compiled are already represented on nixos.org, nix.dev, and Discourse, with varying degrees of structure.
For example, nixos.org shows where to get in touch with the community, nix.dev offers learning materials and links to documentation, and there is a job board on Discourse.
```
